### PR TITLE
Merged Scan and insert functionality

### PR DESCRIPTION
**Pull request by Dingshilun**

### CODE REVIEW

File: src/include/index/skiplist.h [hunk 1]
```diff
@@ -81,7 +95,7 @@ class SkipList {
    */
 
   bool GetFrom(const KeyType &key, const SkipListBaseNode *Node,
```
Comment (DeanChensj) [inline src/include/index/skiplist.h:98]: If we are not using this func, maybe we can remove it

File: src/include/index/skiplist.h [hunk 2]
```diff
@@ -122,15 +136,20 @@ class SkipList {
...
+  NodePair SearchFrom(const KeyType &key, const SkipListBaseNode *Node,
+                      OperationContext &ctx) {
     // TODO: physically deletion when search in the list
     if (Node == nullptr) {
       return std::make_pair(nullptr, nullptr);
     }
     SkipListBaseNode *curr_node = const_cast<SkipListBaseNode *>(Node);
     while (curr_node) {
       SkipListBaseNode *tmp_pointer = curr_node->next_.load();
+
```
Comment (DeanChensj) [inline src/include/index/skiplist.h:149]: We may not want to push this change?

File: src/include/index/skiplist.h [hunk 3]
```diff
@@ -166,34 +185,198 @@ class SkipList {
...
+    LOG_INFO("SearchWithPath %d, levelNow: %u", expected_stored_level,
+             level_now);
     while (level_now >= 0) {
       if (level_now <= expected_stored_level) {
         call_stack[level_now] = SearchFrom(key, curr_node, ctx);
         curr_node = call_stack[level_now].first->down_.load();
       } else {
         curr_node = SearchFrom(key, curr_node, ctx).first->down_.load();
       }
+      // level_now is unsigned
```
Comment (DeanChensj) [inline src/include/index/skiplist.h:205]: Comment should be related to the logic not coding method? 
What about: // Stop at root level

File: src/include/index/skiplist.h [hunk 4]
```diff
@@ -166,34 +185,198 @@ class SkipList {
...
   }
+
+  /*
+ * AddLevel() - add corresponding level to the SkipList
+ *
+ * return true if successfully added or the level is already added
+ * return false if the level cannot be reached from the highest level now
+ */
+  bool AddLevel(u_int32_t level) {
+    LOG_INFO("AddLevel");
```
Comment (DeanChensj) [inline src/include/index/skiplist.h:218]: AddLevel %u,  level?

File: src/include/index/skiplist.h [hunk 5]
```diff
@@ -166,34 +185,198 @@ class SkipList {
...
+          return head->level_ == level;
+        }
+      } else {
+        return true;
+      }
+    }
+  }
+
+  /*
+   * InserNodeIntoInterval() - this method would try to insert the tower into
```
Comment (DeanChensj) [inline src/include/index/skiplist.h:240]: InsertTowerIntoInterval()

File: src/include/index/skiplist.h [hunk 6]
```diff
@@ -166,34 +185,198 @@ class SkipList {
...
+  bool AddLevel(u_int32_t level) {
+    LOG_INFO("AddLevel");
+    SkipListBaseNode *head = this->skip_list_head_.load();
+    if (head->level_ + 1 < level) {
+      return false;
+    } else {
+      if (head->level_ + 1 == level) {
+        SkipListBaseNode *new_head = node_manager_.GetSkipListHead(level);
+        new_head->down_ = head;
+        if (this->skip_list_head_.compare_exchange_strong(head, new_head)) {
```
Comment (DeanChensj) [inline src/include/index/skiplist.h:226]: Don't we need retry here?

File: src/include/index/skiplist.h [hunk 7]
```diff
@@ -166,34 +185,198 @@ class SkipList {
...
+      std::vector<std::pair<SkipListBaseNode *, SkipListBaseNode *>> &
+          call_stack,
+      OperationContext &ctx, u_int32_t start_level = 0,
+      bool check_multiple_key_value=false) {
+    LOG_INFO("InsertTower");
+    u_int32_t expected_level = tower.size();
+    for (u_int32_t i = start_level; i < expected_level; i++) {
+      bool insert_flag = false;
+      do {
+        if (i != 0 && GET_DELETE(tower[i]->GetRoot().load()->next_.load())) {
```
Comment (DeanChensj) [inline src/include/index/skiplist.h:262]: Hmm.. We may want GET_DELETE to work more like GET_NEXT so we don't need to write that long code..

File: src/include/index/skiplist.h [hunk 8]
```diff
@@ -166,34 +185,198 @@ class SkipList {
...
+          }
+          return true;
+        }
+
+        //if the level is 0, multiple key-value pair is required
+        //if the check bool is true, then it should be non-unique index
+        if (i==0){
+          if (!check_multiple_key_value){
+            // unique index, just need to check the next one
+            if (!GET_DELETE(call_stack[i].second->next_.load())
```
Comment (DeanChensj) [inline src/include/index/skiplist.h:276]: clang-format

File: src/include/index/skiplist.h [hunk 9]
```diff
@@ -166,34 +185,198 @@ class SkipList {
...
+                &&ValueCmpEqual(tower[i]->GetValue(),
+                                static_cast<SkipListInnerNode *>(call_stack[i].second)->GetValue())){
+              return false;
+            }
+          }else{
+            //need to check all key-value pairs
```
Comment (DeanChensj) [inline src/include/index/skiplist.h:282]: // There can be node with the same key and different values, we need to ensure no nodes with both key and value are the same

File: src/include/index/skiplist.h [hunk 10]
```diff
@@ -166,34 +185,198 @@ class SkipList {
...
+    LOG_INFO("Insert node");
 
-  bool InsertNode(UNUSED_ATTRIBUTE const KeyType &key,
-                  UNUSED_ATTRIBUTE const ValueType &value,
-                  UNUSED_ATTRIBUTE OperationContext &ctx) {
-    return false;
+    u_int32_t expected_level = 0;
+
+    while (expected_level < max_level_) {
+      if (rand() & 1) {
```
Comment (DeanChensj) [inline src/include/index/skiplist.h:323]: TODO: try not to call too many rand()

File: src/include/index/skiplist.h [hunk 11]
```diff
@@ -166,34 +185,198 @@ class SkipList {
...
+
+    SkipListBaseNode *curr_node = this->skip_list_head_.load();
+
+    while (curr_node->level_ < expected_level) {
+      AddLevel(curr_node->level_ + 1);
+      curr_node = this->skip_list_head_.load();
+    }
+
+    // used to store the path
+    std::vector<std::pair<SkipListBaseNode *, SkipListBaseNode *>> call_stack;
```
Comment (DeanChensj) [inline src/include/index/skiplist.h:338]: std::vector<NodePair>

File: src/include/index/skiplist.h [hunk 12]
```diff
@@ -166,34 +185,198 @@ class SkipList {
...
+    std::vector<SkipListInnerNode *> tower(expected_level + 1);
```
Comment (DeanChensj) [inline src/include/index/skiplist.h:339]: NodeList tower();

File: src/include/index/skiplist.h [hunk 13]
```diff
@@ -166,34 +185,198 @@ class SkipList {
...
+    // else need to verify the next node
+    if (this->duplicate_support_) {
+      // insert the node from the lowest level
+      // redo the search from stack if the insert fails
+      return InsertTowerIntoInterval(key, tower, call_stack, ctx, 0, true);
+    } else {
+      // unique key
+      // need to compare with the second return value's key
+      bool insert_flag;
+      do {
```
Comment (DeanChensj) [inline src/include/index/skiplist.h:360]: I will just assume the logic is correct here.

File: src/include/index/skiplist.h [hunk 14]
```diff
@@ -372,9 +562,115 @@ class SkipList {
...
-  // this one provides the abstract interfaces
-  class SkipListIterator {};
-  class ForwardIterator {};
+  /*
+   * class ForwardIterator - Iterator that supports forward iteration of the
+   * skip list
+   */
+  class ForwardIterator {
+   private:
+    SkipListInnerNode *node_;
```
Comment (DeanChensj) [inline src/include/index/skiplist.h:645]: If you have any suggestion on the elements in this class, please tell me, thanks!

File: src/include/index/skiplist.h [hunk 15]
```diff
@@ -372,9 +562,115 @@ class SkipList {
...
+      node_ = reinterpret_cast<SkipListInnerNode *>(GET_NEXT(node_));
+      if (node_) {
+        kv_p.first = node_->key_;
+        kv_p.second = node_->GetValue();
+      }
+
+      list_->epoch_manager_.LeaveEpoch(epoch_node_p);
+
+      return temp;
+    }
```
Comment (DeanChensj) [inline src/include/index/skiplist.h:745]: Add TODO: more operation need to be override

File: src/include/index/skiplist.h [hunk 16]
```diff
@@ -166,34 +185,198 @@ class SkipList {
...
+   *
+   * It has a call_stack array to accelorate the process
+   * NOTE: this method would retry until the world ends (or the root is
+   *deleted)!!!
+   *
+   * Would only return true, or it would retry until succeed or the root is
+   *deleted
+   * Use this function only the tower can be exactly inserted
+   */
+  bool InsertTowerIntoInterval(
```
Comment (DeanChensj) [inline src/include/index/skiplist.h:251]: move ctx to the last parameter

File: src/include/index/generic_key.h [hunk 17]
```diff
@@ -67,11 +67,10 @@ class GenericKey {
    * @return
    */
   const std::string GetInfo() const {
-    storage::Tuple tuple(schema, data);
+    storage::Tuple tuple(schema, (char *)data);
```
Comment (DeanChensj) [inline src/include/index/generic_key.h:70]: This is the bug in Peleton logging, otherwise we don't want to change this file.

**Review by DeanChensj — COMMENTED** [review]
LGTM, just a few comments.

**Review by DeanChensj — CHANGES_REQUESTED** [review]
Changes requested.

**Review by sxzh93 — APPROVED** [review]

**Pull request merged**